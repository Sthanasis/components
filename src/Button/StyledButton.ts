import styled from "styled-components";
import type { VariantType } from "../../types/types";
import type { ThemeType } from "../../theme";
import opacity from "../assets/opacityHexPrefix";

interface IProps {
  theme: ThemeType;
  variant: VariantType;
  elevated: boolean;
  color: "primary" | "secondary" | "common";
}

export const ButtonContainer = styled.button`
  font-size: ${({ theme }: IProps) => theme.fontSize};
  padding: ${({ theme }: IProps) => theme.padding};
  overflow: hidden;
  cursor: pointer;
  display: flex;
  transition: 0.2s;
  align-items: center;
  justify-content: center;
  position: relative;
  text-decoration: none;
  border: none;
  outline: none;
  background-color: transparent;
  border-radius: ${({ theme }: IProps) => theme.borderRadius};
  ${({ theme, variant, elevated, color }: IProps) => {
    const mainColor = theme.palette[color].main;

    switch (variant) {
      case "contained":
        return `
            background-color: ${mainColor};
            color: #fff;
            &:hover {
                background-color: ${mainColor}${opacity[80]};
            }
            box-shadow: ${elevated ? "0px 2px 4px rgba(0, 0, 0, 0.4)" : "none"};
        `;
      case "outlined":
        return `
          border-width: 1.5px;
          border-style: solid;
          border-color: ${mainColor}${opacity[50]};
          &:hover {
            border-color: ${mainColor};
            background-color: ${mainColor}${opacity[5]};
          }  
        `;
      default:
        return `
            color:${mainColor};
            &:hover {
                background-color: ${mainColor}${opacity[5]};
            }
        `;
    }
  }}
`;
