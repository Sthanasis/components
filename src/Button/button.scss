@import "~/src/css/mixins.scss";

.button {
  @include interactive-element;
  overflow: hidden;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  .icon-start {
    margin-right: 15px;
  }

  .icon-end {
    margin-left: 15px;
  }
  .content {
    position: relative;
  }

  &.fullwidth {
    width: -webkit-fill-available;
  }
  &.primary {
    @include themify("color", "primary");
    &:hover {
      @include themify("background-color", "primary-l");
    }
    span.ripple {
      @include themify("background-color", "primary-m");
    }

    &.contrast {
      color: rgba($contrast-text, 0.8);
      &:hover {
        color: $contrast-text;
      }
      span.ripple {
        background-color: $input-grey-bg;
      }
      &.disabled {
        cursor: not-allowed;
        color: rgba($contrast-text, 0.5);
        &:hover {
          color: rgba($contrast-text, 0.5);
        }
      }
    }

    &.contained {
      @include themify("background-color", "primary");
      color: #fff;
      &:hover {
        @include themify("background-color", "primary-d");
      }
      &.elevated {
        box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.4);
      }
      span.ripple {
        background-color: rgba(255, 255, 255, 0.7);
      }
    }

    &.outlined {
      border-width: 1px;
      border-style: solid;
      @include themify("border-color", "primary-m");
      &:hover {
        @include themify("border-color", "primary");
        @include themify("background-color", "primary-l");
      }
    }
  }
}

@keyframes ripple {
  0% {
    transform: scale(1);
    opacity: 1;
  }
  50% {
    transform: scale(10);
    opacity: 0.375;
  }
  100% {
    transform: scale(35);
    opacity: 0;
  }
}
