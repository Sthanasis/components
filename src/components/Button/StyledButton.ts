import styled from 'styled-components';
import type { VariantType, ButtonType } from '../../types/types';
import type { ThemeType } from '../../../theme';
import opacity from '../../assets/opacityHexPrefix';

interface IProps {
  theme: ThemeType;
  variant: VariantType;
  elevated: boolean;
  buttonType: ButtonType;
}

export const ButtonContainer = styled.button`
  font-size: ${({ theme }: IProps) => theme.fontSize};
  padding: ${({ theme }: IProps) => theme.padding};
  overflow: hidden;
  cursor: pointer;
  display: flex;
  transition: 0.2s;
  align-items: center;
  justify-content: center;
  position: relative;
  text-decoration: none;
  border: none;
  outline: none;
  background-color: transparent;
  border-radius: ${({ theme }: IProps) => theme.borderRadius};
  ${({ theme, variant, elevated, buttonType }: IProps) => {
    const mainColor = theme.palette[variant].main;
    switch (buttonType) {
      case 'contained':
        return `
            background-color: ${mainColor};
            color: #fff;
            &:hover {
                background-color: ${mainColor}${opacity[80]};
            }
            &:disabled {
                background-color: gray;
                color: lightgray;
                cursor: not-allowed;
            }
            box-shadow: ${elevated ? '0px 2px 4px rgba(0, 0, 0, 0.4)' : 'none'};
        `;
      case 'outlined':
        return `
          color:${mainColor};
          border-width: 1.5px;
          border-style: solid;
          border-color: ${mainColor}${opacity[50]};
          &:hover {
            border-color: ${mainColor};
            background-color: ${mainColor}${opacity[5]};
          }
          &:disabled {
            color: lightgray;
            border-color: lightgray;
            cursor: not-allowed;
            background-color: initial;
          }  
        `;
      default:
        return `
            color:${mainColor};
            &:hover {
                background-color: ${mainColor}${opacity[5]};
            }
            &:disabled {
                color: lightgray;
                cursor: not-allowed;
                background-color: initial;
            }
        `;
    }
  }}
`;
