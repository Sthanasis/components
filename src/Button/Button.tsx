import { MouseEvent, ReactNode, memo } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { useTheme } from 'styled-components';
import { IBaseProps } from '../../types/props';
import { ButtonContainer } from './StyledButton';
import Ripple from '../Rippler/Ripple';
import type { VariantType, ButtonType } from '../../types/types';
export interface IButtonProps extends IBaseProps {
  children?: React.ReactNode | React.ReactNode[];
  buttonType?: ButtonType;
  variant?: VariantType;
  elevated?: boolean;
  onClick?: (...args: unknown[]) => unknown;
  fullwidth?: boolean;
  contrast?: boolean;
  disabled?: boolean;
  iconStart?: IconProp;
  iconEnd?: IconProp;
  icon?: IconProp;
  showOnlyIcons?: boolean;
  disableTheme?: boolean;
}

const Button = ({
  style,
  className,
  children,
  variant = 'primary',
  buttonType = 'text',
  elevated = true,
  onClick,
  fullwidth = false,
  contrast = false,
  disabled = false,
  iconStart,
  iconEnd,
  testId,
  icon,
  showOnlyIcons,
}: IButtonProps): JSX.Element => {
  const theme = useTheme();

  const click = (e: MouseEvent<HTMLButtonElement>) => {
    if (onClick) {
      onClick(e);
    }
  };

  let childNodes: ReactNode = (
    <>
      {iconStart && (
        <FontAwesomeIcon className="icon icon-start" icon={iconStart} />
      )}
      {children}
      {iconEnd && <FontAwesomeIcon className="icon icon-end" icon={iconEnd} />}
    </>
  );

  if (icon) {
    childNodes = <FontAwesomeIcon icon={icon} />;
  } else if (showOnlyIcons) {
    childNodes = (
      <>
        {iconStart && (
          <FontAwesomeIcon className="icon icon-start" icon={iconStart} />
        )}
        {iconEnd && (
          <FontAwesomeIcon className="icon icon-end" icon={iconEnd} />
        )}
      </>
    );
  }

  return (
    <ButtonContainer
      className={className}
      onClick={click}
      disabled={disabled}
      data-testid={testId}
      buttonType={buttonType}
      theme={theme}
      variant={variant}
      elevated={elevated}
    >
      <div className="content">{childNodes}</div>
      <Ripple disabled={disabled} />
    </ButtonContainer>
  );
};

export default memo(Button);
