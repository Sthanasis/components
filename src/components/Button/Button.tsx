import { MouseEvent, ReactNode, memo } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { useTheme } from 'styled-components';
import { IBaseProps } from 'src/types/props';
import { ButtonContainer } from './StyledButton';
import Ripple from '../Rippler/Ripple';
import type { VariantType, ButtonType } from 'src/types/types';

export interface IButtonProps extends IBaseProps {
  children?: React.ReactNode | React.ReactNode[];
  buttonType?: ButtonType;
  variant?: VariantType;
  elevated?: boolean;
  onClick?: (...args: unknown[]) => unknown;
  fullwidth?: boolean;
  contrast?: boolean;
  disabled?: boolean;
  iconStart?: IconProp;
  iconEnd?: IconProp;
  icon?: IconProp;
}

const Button = ({
  className,
  children,
  variant = 'primary',
  buttonType = 'text',
  elevated = true,
  onClick,
  fullwidth = false,
  contrast = false,
  disabled = false,
  iconStart,
  iconEnd,
  testId,
  icon,
}: IButtonProps): JSX.Element => {
  const theme = useTheme();

  const click = (e: MouseEvent<HTMLButtonElement>) => {
    if (onClick) {
      onClick(e);
    }
  };

  let childNodes: ReactNode = (
    <>
      {iconStart && (
        <FontAwesomeIcon className="icon icon-start" icon={iconStart} />
      )}
      {children}
      {iconEnd && <FontAwesomeIcon className="icon icon-end" icon={iconEnd} />}
    </>
  );

  if (icon) {
    childNodes = <FontAwesomeIcon icon={icon} />;
  }

  return (
    <ButtonContainer
      className={className}
      onClick={click}
      disabled={disabled}
      data-testid={testId}
      buttonType={buttonType}
      theme={theme}
      variant={variant}
      elevated={elevated}
    >
      <div className="content">{childNodes}</div>
      <Ripple disabled={disabled} />
    </ButtonContainer>
  );
};

export default memo(Button);
