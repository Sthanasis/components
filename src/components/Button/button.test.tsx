import { render, screen, fireEvent } from '@testing-library/react';
import Button from '.';
import { fa0 } from '@fortawesome/free-solid-svg-icons';

describe('Button Component', () => {
  it('renders an empty button', () => {
    const { queryByRole } = render(<Button />);
    expect(queryByRole('button')).toMatchSnapshot();
  });

  it('renders the children prop', () => {
    const { getByText } = render(<Button>TEST</Button>);
    expect(getByText('TEST')).toBeInTheDocument();
  });

  it('renders a button with red text when passed through style prop', () => {
    const { queryByRole } = render(
      <Button style={{ color: 'red' }}>TEST</Button>
    );
    expect(queryByRole('button')).toHaveAttribute('style', 'color: red;');
  });
});
