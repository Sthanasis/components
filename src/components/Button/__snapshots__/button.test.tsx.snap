// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders an empty button 1`] = `
<button
  class="sc-bcXHqe knhseT"
>
  <div />
  <div
    class="sc-gswNZR jFVfHn"
    color="#fff"
  />
</button>
`;
